src/main/scala/org/apache/spark/sql/rikai/Image.scala

```diff
@@ -28,21 +28,37 @@ private[spark] class ImageType extends UserDefinedType[Image] {
   override def sqlType: DataType =
     StructType(
       Seq(
-        StructField("uri", StringType, false)
+        StructField("data", BinaryType, true),
+        StructField("uri", StringType, true)
       )
     )
 
   override def pyUDT: String = "rikai.spark.types.vision.ImageType"
 
   override def serialize(obj: Image): Any = {
-    val row = new GenericInternalRow(1);
-    row.update(0, UTF8String.fromString(obj.uri))
+    val row = new GenericInternalRow(2);
+    if (obj.data.isDefined) {
+      row.update(0, obj.data.get)
+    }
+    if (obj.uri.isDefined) {
+      row.update(1, UTF8String.fromString(obj.uri.get))
+    }
     row
   }
 
   override def deserialize(datum: Any): Image =
     datum match {
-      case row: InternalRow => new Image(row.getString(0).toString())
+      case row: InternalRow =>
+        new Image(
+          row.isNullAt(0) match {
+            case true  => None
+            case false => Some(row.getBinary(0))
+          },
+          row.isNullAt(1) match {
+            case true  => None
+            case false => Some(row.getString(1).toString)
+          }
+        )
```

Comment:
this will throw if it's not an InternalRow.

Reply:
It seems wont happen? this is merely used for doing the type conversion., for example, `spark.ml` does the same

https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/linalg/VectorUDT.scala#L54-L69

Reply:
for scala, exhaustive pattern matching is considered best practice. In 2.13.4 for example strict exhaustive pattern matching is on by default. In older versions it's just a warning.

Reply:
What exception do you suggest to throw here?

Reply:
no strong opinion here. Maybe IllegalArgumentException?